### PR TITLE
Feature/222 permissions

### PR DESCRIPTION
**Pull request by amylo**
I've implemented the base permissions system. This PR is for the following permissions:
1. Public
2. Authenticated

The RoutingService will be in charge of answering the question of whether the app will navigate to a given page, but checking with current state against the permissions required to access the page.

After this PR, it should be fairly straight forward to implement permission requirements for the rest of our routes, eg. for organizers, volunteers, etc.

### CODE REVIEW

File: src/app/routing/services/RoutingService.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import { isEmpty, isLoaded } from "react-redux-firebase";
+import routes, { IRoute, IRoutes } from "../routes";
+import ROUTE_PERMISSIONS from "../RoutePermissionsList";
+import { IRoutePermission } from "./utils";
+import { PERMISSIONS } from "../../model";
+
+export class RoutingService {
+  private _auth: AuthState = {};
```
[the remark this answers is not in the document]
Comment (mat10tng) [reply src/app/routing/services/RoutingService.ts:8]: woa woa, Paul the progressive js man. Did not even know about this, TIL

File: src/app/routing/RoutePermissionsList.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  routes.missions.feedback,
+  routes.missions.details,
+  routes.missions.main,
+  routes.organizer.dashboard.home,
+  routes.organizer.dashboard.missions,
+  routes.organizer.dashboard.recipients,
+  routes.organizer.dashboard.volunteer,
+  routes.request.foodbox,
+  routes.request.success,
+  routes.request.error,
```
[the remark this answers is not in the document]
Comment (amylo) [reply src/app/routing/RoutePermissionsList.ts:31]: oh interesting, sure i can make em public

File: src/app/routing/services/utils.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+export interface IRoutePermission {
+  route: IRoute;
+  requiredPermissions: PERMISSIONS[];
+}
+/**
+ * Add the same permission to multiple routes.
+ */
+export function addPermissionToRoutes(
```
[the remark this answers is not in the document]
Comment (amylo) [reply src/app/routing/services/utils.ts:28]: i wrote 3 different functions that do the same thing actually, I wasn't sure which I wanted to keep LOL:

1. add same permission to multiple routes (handy for setting AUTHENTICATED for # routes)
2. add multiple permissions to same route (handy if we want to change the way we write permissions to more route-based)
3. add permissions to different routes (handy when i want to assign multiple permissions to each route in one call)

File: src/app/routing/services/RoutingService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (amylo) [reply src/app/routing/services/RoutingService.ts:8]: TypeScript had public/private for ages, i love to use it

**Comment by amylo** [comment]
> This is awesome.! So much easier to handle these permissions.
> 
> One thing I thought I'd run by you is if you considered using `map` instead for the `RoutePermissionList`?
> Something like this
> 
> ```ts
> interface IRoutPermissions{
>   [key: IRoute] : Set<PERMISSIONS>
> }
> ```
> 
> so each route string maps to a `Set` of permissions which would make it super easy for grabbing and setting permissions. Not to mention not having to loop over an array to find the route each time

That i so true. I think I was thinking this eventually could be defined in the DB rather in the code. And my mind went straight to Arrays, but I can refactor this to simply be Map :)

**Pull request merged**